specification/hugr.md

```diff
@@ -1170,44 +1160,71 @@ The new hugr is then derived as follows:
+A _partial hugr_ is a graph formed by a subset of nodes of a valid hugr together
+with a subset of their adjoining edges. It must not include a `Module` node.
 
-  - a partial hugr G;
+Informally, a partial hugr is is a graph $G$ having the same node and edge types as a hugr
+and satisfying all the same local constraints except that it may be missing edges.
 
-  - a map T from top(G) to IDs of container nodes in Γ\\S\*;
+Given a partial hugr $G$, let
 
-  - a map B from bot(G) to IDs of container nodes in S\*, such that B(x)
-    is separated from B(y) unless x == y. Let X be the set of children
-    of values in B, and R be S\*\\X\*.
+  - $\top(G)$ be the set of nodes in $G$ without an incoming hierarchy edge;
+  - $\bot(G)$ be the set of container nodes in $G$ without an outgoing hierarchy edge.
 
-  - a bijection μ<sub>inp</sub> between inp(G) and the set of input
-    ports of nodes in R whose source is not in R;
+Given a set $S$ of nodes in a hugr, let $S^\*$ be the set of all nodes
+descended from nodes in $S$ (i.e. reachable from $S$ by following hierarchy edges),
+including $S$ itself.
 
-  - a bijection μ<sub>out</sub> between out(G) and the set of output
-    ports of nodes in R whose target is not in R;
+Call two nodes $a, b \in \Gamma$ _separated_ if $a \notin \\{b\\}^\*$ and
+$b \notin \\{a\\}^\*$ (i.e. there is no hierarchy relation between them).
 
-  - disjoint subsets Init and Term of top(G);
+We a structure to specify edge insertion.
 
-The new hugr is then derived by:
+A `NewEdgeSpec` specifies an edge inserted between an existing node and a new node.
+It contains the following fields:
 
-1.  adding the new nodes from G;
+  - `SrcNode`: the source node of the new edge.
+  - `TgtNode`: the target node of the new edge.
+  - `EdgeKind`: may be `Value`, `Order`, `Static` or `ControlFlow`.
+  - `SrcPos`: for `Value` and `Static` edges, the position of the source port;
+    for `ControlFlow` edges, the position among the outgoing edges.
+  - `TgtPos`: (for `Value` and `Static` edges only) the desired position among
+    the incoming ports to the new node.
 
-2.  connecting the ports according to the bijections μ<sub>inp</sub> and
-    μ<sub>out</sub>;
+Note that in a `NewEdgeSpec` one of `SrcNode` and `TgtNode` is an existing node
+in the hugr and the other is a new node.
 
-3.  for each node n in top(G), adding a hierarchy edge from t(n) to n,
-    placing n in the first position among children of t(n) if n is in
-    Init and in the second position if n is in Term;
+The `Replace` method takes as input:
 
-4.  for each node n in bot(G), and for each child m of b(n), adding a
-    hierarchy edge from n to m (replacing m’s existing parent edge)
+  - a set $S$ of mutually-separated nodes in $\Gamma$;
+  - a partial hugr $G$;
+  - a map $T : \top(G) \to \Gamma \setminus S^*$ whose image consists of container nodes;
+  - a map $B : \bot(G) \to S^\*$ whose image consists of container nodes, such that $B(x)$
```

Comment:
Yes they are, but we don't need to say it here because it's part of the definition of bot(G).

Wikipedia says they are synonyms but that "range" should be avoided because it's also sometimes used to mean the codomain (so they'd only be the same for surjective functions).